pkg/integrations/v2/eventhandler/eventhandler.go

```diff
@@ -0,0 +1,454 @@
+	var (
+		config *rest.Config
+		err    error
+	)
+
+	if c.InCluster {
+		config, err = rest.InClusterConfig()
+	} else {
+		config, err = clientcmd.BuildConfigFromFlags("", c.KubeconfigPath)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// get an informer
+	factory := informers.NewSharedInformerFactory(clientset, time.Duration(c.InformerResync)*time.Second)
+	eventInformer := factory.Core().V1().Events().Informer()
+
+	eh := &EventHandler{
+		LokiClient:    globals.Logs.Instance(c.LogsInstance),
+		Log:           l,
+		CachePath:     c.CachePath,
+		EventInformer: eventInformer,
+		ClusterLabel:  c.ClusterName,
+		SendTimeout:   time.Duration(c.SendTimeout) * time.Second,
+	}
+	// set the resource handler fns
+	eh.initInformer(eventInformer)
+	return eh, nil
+}
+
+// Initialize informer by setting event handler fns
+func (eh *EventHandler) initInformer(eventsInformer cache.SharedIndexInformer) {
+	eventsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    eh.addEvent,
+		UpdateFunc: eh.updateEvent,
+		DeleteFunc: eh.deleteEvent,
+	})
+}
+
+// Handles new event objects
+func (eh *EventHandler) addEvent(obj interface{}) {
+	e, ok := obj.(*v1.Event)
+	if !ok {
+		level.Error(eh.Log).Log("msg", "Object not v1.Event", "obj", obj)
+		return
+	}
+	eventTs := getTimestamp(e)
+
+	// if event is older than the one stored in cache on startup, we've shipped it
+	if eventTs.Before(eh.InitEvent.Timestamp) {
+		return
+	}
+	// if event is equal and is in map, we've shipped it
+	if eventTs.Equal(eh.InitEvent.Timestamp) {
+		if _, ok := eh.InitEvent.RvMap[e.ResourceVersion]; ok {
+			return
+		}
+	}
+
+	labels, msg, err := eh.handleEvent(e)
+	if err != nil {
+		level.Error(eh.Log).Log("msg", "Error processing event", "err", err, "event", e)
+	}
+
+	entry := newEntry(msg, eventTs, labels)
+	ok = eh.LokiClient.SendEntry(entry, eh.SendTimeout)
+	if !ok {
+		err = fmt.Errorf("error handing entry off to promtail")
+		level.Error(eh.Log).Log("err", err, "entry", entry)
+		return
+	}
+	level.Info(eh.Log).Log("msg", "Shipped entry", "eventRV", e.ResourceVersion, "eventMsg", e.Message)
+
+	// update cache with new "last" event
+	err = eh.updateLastEvent(e, eventTs)
+	if err != nil {
+		level.Error(eh.Log).Log("msg", "Could not write latest event to cache", "err", err)
+	}
+}
+
+// Handles event object updates. Note that this get triggered on informer resyncs and also
+// events occurring more than once (in which case .count is incremented)
+func (eh *EventHandler) updateEvent(objOld interface{}, objNew interface{}) {
+	eOld, ok := objOld.(*v1.Event)
+	if !ok {
+		level.Error(eh.Log).Log("msg", "Object not v1.Event", "obj", eOld)
+		return
+	}
+
+	eNew, ok := objNew.(*v1.Event)
+	if !ok {
+		level.Error(eh.Log).Log("msg", "Object not v1.Event", "obj", eNew)
+		return
+	}
+
+	if eOld.GetResourceVersion() == eNew.GetResourceVersion() {
```

Comment:
Just curious: what's a resync update? When would this happen?

Reply:
Every "resync interval" (one of the parameters passed in when creating an informer, and consequently also the integration config), the informer will basically do a relist and sync its local cache with the control plane / etcd. When it does this relist, it triggers update events for all objects in the cache, so we check to see if the object has changed at all (new RV). If it's just the same old event that we already shipped, we ignore. See [9. here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md)